### DISCUSSION TITLE
GPU comutations

### DISCUSSION THREAD

**Discussion by hosseinrezaa** [InteractiveComputerGraphics/SPlisHSPlasH#164] [Q&A]
Hi
I installed the splishsplash on a machine with Quadro p5000 GPU. But when I am running te code it seems that GPU is not active.
What should I do to perform the calculations with GPU.
![Untitled](https://user-images.githubusercontent.com/48132503/134502526-27bf4cf3-3ef1-4f29-be42-52c17bca47c1.jpg)
Regards

**Comment by janbender** [comment]
You have to activate the CMake option "USE_GPU_NEIGHBORHOODSEARCH".

**Comment by hosseinrezaa** [comment]
I have activated the option "USE_GPU_NEIGHBORHOODSEARCH". But when I run the program still cannot run by GPU. Is there any adjustment in Visual Studio? or windows?

**Comment by janbender** [comment]
Did you install Cuda?

**Comment by hosseinrezaa** [comment]
Yes, Cuda toolkit 10.1

**Comment by janbender** [comment]
Do you see a message "Use cuNSearch for neighborhood search" when you execute CMake?

**Comment by hosseinrezaa** [comment]
Yes. I use cmake and I see this message. But when I try to run the C++ program using Visual studio only it uses the CPU .

If your CPU does not support AVX, turn off the USE_AVX flag.
Using Win32 for window creation
Use cuNSearch for neighborhood search
The CUDA compiler identification is NVIDIA 11.4.120
Detecting CUDA compiler ABI info
Detecting CUDA compiler ABI info - done
Check for working CUDA compiler: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.4/bin/nvcc.exe - skipped
Detecting CUDA compile features
Detecting CUDA compile features - done
Found CUDA: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.4 (found suitable version "11.4", minimum required is "9.0") 
Configuring done
Generating done

**Comment by janbender** [comment]
That is not possible since cuNSearch is used which does not implement any CPU computation.

**Comment by janbender** [comment]
Note that most of the work is done in the pressure solver which is not implemented on the GPU. There is one pull request with a Cuda implementation. However, in our tests this was slower than avx.

**Comment by hosseinrezaa** [comment]
I am now running the code

![image](https://user-images.githubusercontent.com/48132503/135052188-727e8fb3-e26a-4baa-bafd-f642f46cc2b5.png)

**Comment by hosseinrezaa** [comment]
So it is using the GPU and if I need faster runs should I try with better devices?

**Comment by janbender** [comment]
The most time consuming part is typically the pressure solver. The fastest one is DFSPH with AVX support. As far as I can see you are also using an implicit viscosity solver. This means that another linear system has to be solved which is also time consuming. 

More CPU cores will help to make the simulation faster.

**Comment by hosseinrezaa** [comment]
Thank you so much

**Comment by nobane** [comment]
The correct option to pass to cmake is `-DUSE_GPU_NEIGHBORHOOD_SEARCH=On`.

**Comment by chunleili** [comment]
> I am now running the code
> 
> ![image](https://user-images.githubusercontent.com/48132503/135052188-727e8fb3-e26a-4baa-bafd-f642f46cc2b5.png)
> 

I have the same problem with you. The usage of GPU is really low, so I dont know whether the GPU is used. And I dont feel the simulation become faster.

**Reply by woshitx** [reply]
你好，请问你是怎么运行成功的呢，只需要启用-DUSE_GPU_NEIGHBORHOOD_SEARCH=On.选项就行了吗，还需要改其他地方吗。

**Comment by janbender** [comment]
As I already mentioned, currently only the neighborhood search is implemented on GPU. From exchanging an optimized neighborhood search on CPU with the GPU version you cannot expect a incredible speedup of the overall simulation since the neighborhood search is not the most time-consuming part of the simulation. Moreover, in small simulations (I see 57k particles in the screenshot) using the GPU does not bring a big speed up. Hope that in future we will also find the time to reimplement the pressure solver on the GPU.

**Comment by woshitx** [comment]
Hello, I encountered no errors during the cmake process, and I could see the message 'Use cuNSearch for neighborhood search', but I received an error during the make process. Apart from the option -DUSE_GPU_NEIGHBORHOOD_SEARCH=On, do I need to modify anything else?

**Comment by janbender** [comment]
You need a cuda installation.

**Reply by woshitx** [reply]
![微信图片_20240918200415](https://github.com/user-attachments/assets/e2f350c4-39c4-4c2f-b55e-3e4349a379cc)
I have downloaded CUDA
